miner/env_changes.go

```diff
@@ -0,0 +1,417 @@
+		}
+	}
+
+	cfg := *chData.chain.GetVMConfig()
+	touchTracer := logger.NewAccountTouchTracer()
+	cfg.Tracer = touchTracer
+	cfg.Debug = true
+
+	c.env.state.SetTxContext(tx.Hash(), c.env.tcount+len(c.txs))
+	receipt, err := core.ApplyTransaction(chData.chainConfig, chData.chain, &c.env.coinbase, c.gasPool, c.env.state, c.env.header, tx, &c.usedGas, cfg, nil)
+	if err != nil {
+		c.rollback(usedGasBefore, gasPoolBefore, profitBefore, txsBefore, receiptsBefore)
+
+		switch {
+		case errors.Is(err, core.ErrGasLimitReached):
+			// Pop the current out-of-gas transaction without shifting in the next from the account
+			from, _ := types.Sender(signer, tx)
+			log.Trace("Gas limit exceeded for current block", "sender", from)
+			return receipt, popTx, err
+
+		case errors.Is(err, core.ErrNonceTooLow):
+			// New head notification data race between the transaction pool and miner, shift
+			from, _ := types.Sender(signer, tx)
+			log.Trace("Skipping transaction with low nonce", "sender", from, "nonce", tx.Nonce())
+			return receipt, shiftTx, err
+
+		case errors.Is(err, core.ErrNonceTooHigh):
+			// Reorg notification data race between the transaction pool and miner, skip account =
+			from, _ := types.Sender(signer, tx)
+			log.Trace("Skipping account with hight nonce", "sender", from, "nonce", tx.Nonce())
+			return receipt, popTx, err
+
+		case errors.Is(err, core.ErrTxTypeNotSupported):
+			// Pop the unsupported transaction without shifting in the next from the account
+			from, _ := types.Sender(signer, tx)
+			log.Trace("Skipping unsupported transaction type", "sender", from, "type", tx.Type())
+			return receipt, popTx, err
+
+		default:
+			// Strange error, discard the transaction and get the next in line (note, the
+			// nonce-too-high clause will prevent us from executing in vain).
+			log.Trace("Transaction failed, account skipped", "hash", tx.Hash(), "err", err)
+			return receipt, shiftTx, err
+		}
+	}
+
+	for _, address := range touchTracer.TouchedAddresses() {
+		if _, in := chData.blacklist[address]; in {
+			c.rollback(usedGasBefore, gasPoolBefore, profitBefore, txsBefore, receiptsBefore)
+			return nil, popTx, errors.New("blacklist violation, tx trace")
+		}
+	}
+
+	c.profit.Add(c.profit, new(big.Int).Mul(new(big.Int).SetUint64(receipt.GasUsed), gasPrice))
+	c.txs = append(c.txs, tx)
+	c.receipts = append(c.receipts, receipt)
+
+	return receipt, shiftTx, nil
+}
+
+func (c *envChanges) commitBundle(bundle *types.SimulatedBundle, chData chainData) error {
+	var (
+		profitBefore   = new(big.Int).Set(c.profit)
+		coinbaseBefore = new(big.Int).Set(c.env.state.GetBalance(c.env.coinbase))
+		gasUsedBefore  = c.usedGas
+		gasPoolBefore  = new(core.GasPool).AddGas(c.gasPool.Gas())
+		txsBefore      = c.txs[:]
+		receiptsBefore = c.receipts[:]
+		hasBaseFee     = c.env.header.BaseFee != nil
+
+		bundleErr error
+	)
+
+	for _, tx := range bundle.OriginalBundle.Txs {
+		if hasBaseFee && tx.Type() == types.DynamicFeeTxType {
```

Comment:
I hate this part of our bundle commit code. I've added this to the first version of our builder as an inheritance from other code that worked with bundles but this is just out of place tbh. This is out of scope for this pr. We can probably do that in separate one. 

Normally, for transactions these checks are done in txpool so when tx is in the `miner` its already sanity checked. But for some reason we do that for bundles here instead of doing that in txpool. 

I did this for sbundles, they are validated in txpool. I think we should do the same thing for bundles and make code for them smaller.

Reply:
I've added TODO to move this to txpool, it makes sense to move the logic there and would help make this loop more compact while putting validation logic in one place